docs/code/IdvFileFormat.md

```diff
@@ -0,0 +1,191 @@
+in one vector valued column.)
+
+The actual values are stored in blocks. A block holds values for a single
+column across multiple rows. Block format is dictated by a codec. There is a
+table-of-contents and lookup table to facilitate quasi-random access to
+particular blocks. (Quasi in the sense that looking up the value for a column
+and particular row may require .)
+
+## General Data Format
+
+Before we discuss the format itself we will establish some conventions on how
+individual scalar values, strings, and other data is serialized. All basic
+pieces of data (e.g., a single number, or a single string) are encoded in ways
+reflecting the semantics of the .NET `BinaryWriter` class, those semantics
+being:
+
+* All numbers are stored as little-endian, using their natural fix-length
+  binary encoding.
+
+* Strings are stored using an unsigned LEB128 number describing the number of
+  bytes, followed by that many bytes containing the UTF-8 encoded string.
+
+A note about this: [LEB128](https://en.wikipedia.org/wiki/LEB128) is a simple
+encoding to encode arbitrarily large integers. Each byte of 8-bits follows
+this convention. The most significant bit is 0 if and only if this is the end
+of the LEB128 encoding. The remaining 7 bits are a part of the number being
+encoded. The bytes are stored little-endian, that is, the first byte holds the
+7 least significant bits, the second byte (if applicable) holds the next 7
+least significant bits, etc., and the last byte holds the 7 most significant
+bits. LEB128 is used one or two places in this format. (I might tend to prefer
+use of LEB128 in places where we are writing values that, on balance, we
+expect to be relatively small, and only in cases where there is no potential
+for benefit for random access to the associated stream, since LEB128 is
+incompatible with random access. However this is not formulated into anything
+approaching a definite policy.)
+
+## Header
+
+Every binary instances stream has a header composed of 256 bytes, at the start
+of the stream. Not all bytes are used. Those bytes that are not explicitly
+used have undefined content, and can have anything in them. We strongly
+encourage writers of this format to insert obscene messages in this dead
+space. The content is defined as follows (the offsets being the start of that
+column).
+
+Offsets | Type  | Name and Description
+--------|-------|---------------------
+0       | ulong | **Signature**: The magic number of this file.
+8       | ulong | **Version**: Indicates the version of the data file.
+16      | ulong | **CompatibleVersion**: Indicates the minimum reader version that can interpret this file, possibly with some data loss.
+24      | long  | **TableOfContentsOffset**: The offset to the column table of contents structure.
+32      | long  | **TailOffset**: The eight-byte tail signature starts at this offset. So, the entire dataset stream should be considered to have eight plus this value bytes.
```

Comment:
"this value bytes" should be (something like) "this value in bytes"

Reply:
This was actually grammatically correct but, I agree, slightly confusing. I think you were parsing it as "(eight) plus (this value bytes)", but the correct interpretation is "(eight plus this value) bytes". I have changed it though, to say it has "byte length of eight plus this value," to remove any misunderstandings.

---
In reply to: [188814633](https://github.com/dotnet/machinelearning/pull/173#discussion_r188814633) [](ancestors = 188814633)